packages/variant-theme-v9/src/ThemeGenerator.ts

```diff
@@ -0,0 +1,428 @@
+  ThemeGenerator,
+  themeRulesStandardCreator,
+  updateA
+} from '@fluentui/react';
+import { BrandVariants, Theme } from '@fluentui/react-components';
+import { VariantThemeType } from '@fluentui/scheme-utilities';
+import {
+  black,
+  white,
+  grey,
+  whiteAlpha,
+  blackAlpha,
+  AlphaColors,
+  type Greys
+} from '@fluentui/react-migration-v8-v9';
+import * as lodash from 'lodash';
+import { contrast, hex_to_sRGB } from './ThemeDesigner/src/colors';
+
+const standardContrastRatio: number = 4.5;
+const whiteHex: string = '#ffffff';
+const blackHex: string = '#000000';
+const greyReverse: Record<string, Greys> = Object.keys(grey).reduce((acc: any, key: any) => {
+  acc[grey[key]] = key;
+  return acc;
+}, {});
+
+export const getV8Theme = (primaryColor: string, darkMode = false) => {
+  const themeRules = themeRulesStandardCreator();
+
+  ThemeGenerator.setSlot(
+    themeRules[BaseSlots[BaseSlots.foregroundColor]],
+    darkMode ? whiteHex : '#323130',
+    darkMode,
+    false,
+    darkMode
+  );
+  ThemeGenerator.setSlot(
+    themeRules[BaseSlots[BaseSlots.backgroundColor]],
+    darkMode ? '#323130' : whiteHex,
+    darkMode,
+    false,
+    darkMode
+  );
+  ThemeGenerator.setSlot(themeRules[BaseSlots[BaseSlots.primaryColor]], primaryColor, darkMode, true, true);
+
+  const themeAsJson = ThemeGenerator.getThemeAsJson(themeRules);
+
+  const finalTheme = createTheme({
+    ...{ palette: themeAsJson },
+    isInverted: darkMode || isDark(themeRules[BaseSlots[BaseSlots.backgroundColor]].color!)
+  });
+
+  return finalTheme;
+};
+
+export function getVariantTheme(
+  brandVariants: BrandVariants,
+  theme: Theme,
+  type: VariantThemeType,
+  isInverted: boolean
+) {
+  // Need a way to detect custom neutrals in the theme
+  // Need a way to detect if the theme is light or dark
+
+  const newTheme: Theme = {
+    ...theme
+  };
+  switch (type) {
+    case VariantThemeType.Neutral:
+      // Set the background to be a light grey
+      // Shift all the other neutrals to account for the darker background
+      getNeutralVariant(newTheme, isInverted);
+      return newTheme;
+    case VariantThemeType.Soft:
+      // Set the background to be a light tint of the brand color
+      // Shift all the other neutrals to account for the darker background
+      getSoftVariant(newTheme, brandVariants, isInverted);
+      return newTheme;
+    case VariantThemeType.Strong:
+      getStrongVariant(newTheme, brandVariants, isInverted);
+      return newTheme;
+    default:
+      return newTheme;
+  }
+}
+
+function getNeutralVariant(theme: Theme, isInverted: boolean) {
+  for (const token in theme) {
+    if (token.startsWith('colorNeutralBackground') && !isTokenNameContains(token, ['Alpha'])) {
+      const num: number | undefined = getNumberFromToken(token);
+      if (!num) {
+        continue;
+      }
+      if (num >= 4 && num <= 5) {
+        if (token.endsWith('hover')) {
+          theme[token] = shift(theme[token], !isInverted, 12);
```

Comment:
this reminds me that we should put some high level comments at the top to describe the general design principles.